src/core/algorithms/nar/des/encoded_nar.cpp

```diff
@@ -0,0 +1,101 @@
+#include "encoded_nar.h"
+
+#include "algorithms/nar/value_range.h"
+#include "model/types/types.h"
+
+namespace algos::des {
+
+using model::NAR;
+
+size_t EncodedNAR::VectorSize() const {
+    return encoded_value_ranges_.size() * EncodedValueRange::kFieldCount + 1;
+}
+
+size_t EncodedNAR::FeatureCount() const {
+    return encoded_value_ranges_.size();
+}
+
+// TODO: remove code duplication here
+double& EncodedNAR::operator[](size_t index) {
+    qualities_consistent_ = false;
+    if (index == 0) {
+        return implication_sign_pos_;
+    }
+    index--;
+    size_t feature = index / EncodedValueRange::kFieldCount;
+    size_t feature_field = index % EncodedValueRange::kFieldCount;
+    return encoded_value_ranges_[feature][feature_field];
+}
+
+double const& EncodedNAR::operator[](size_t index) const {
+    if (index == 0) {
+        return implication_sign_pos_;
+    } else {
+        index--;
+        size_t feature = index / EncodedValueRange().kFieldCount;
+        size_t feature_field = index % EncodedValueRange().kFieldCount;
+        return encoded_value_ranges_[feature][feature_field];
+    }
+}
+
+NAR EncodedNAR::SetQualities(FeatureDomains& domains, TypedRelation const* typed_relation) {
+    NAR this_decoded = Decode(domains);
+    this_decoded.SetQualities(typed_relation);
+    qualities_ = this_decoded.GetQualities();
+    qualities_consistent_ = true;
+    return this_decoded;
+}
+
+model::NARQualities const& EncodedNAR::GetQualities() const {
+    if (!qualities_consistent_) {
+        throw std::logic_error("Getting uninitialized qualities from NAR.");
+    }
+    return qualities_;
+}
+
+NAR EncodedNAR::Decode(FeatureDomains& domains) const {
```

Comment:
```suggestion
NAR EncodedNAR::Decode(FeatureDomains const& domains) const {
```
If `FeatureDomains` doesn't have `const operator[]` (some of the stl containers may have this problem), use `at`